src/commands.ts

```diff
@@ -383,36 +387,39 @@ async function openWorkspace(
     newWindow = false
   }
 
-  // If a folder isn't specified, we can try to open a recently opened folder.
-  if (!folderPath) {
+  // If a folder isn't specified or we have been asked to open the most recent,
+  // we can try to open a recently opened folder/workspace.
+  if (!folderPath || openRecent) {
     const output: {
       workspaces: { folderUri: vscode.Uri; remoteAuthority: string }[]
     } = await vscode.commands.executeCommand("_workbench.getRecentlyOpened")
     const opened = output.workspaces.filter(
-      // Filter out `/` since that's added below.
+      // Remove recents that do not belong to this connection.  The remote
+      // authority maps to a workspace or workspace/agent combination (using the
+      // SSH host name).  This means, at the moment, you can have a different
+      // set of recents for a workspace versus workspace/agent combination, even
+      // if that agent is the default for the workspace.
       (opened) => opened.folderUri?.authority === remoteAuthority,
     )
-    if (opened.length > 0) {
-      let selected: (typeof opened)[0]
 
-      if (opened.length > 1) {
-        const items: vscode.QuickPickItem[] = opened.map((folder): vscode.QuickPickItem => {
-          return {
-            label: folder.folderUri.path,
-          }
-        })
-        const item = await vscode.window.showQuickPick(items, {
-          title: "Select a recently opened folder",
-        })
-        if (!item) {
-          return
+    // openRecent will always use the most recent.  Otherwise, if there are
+    // multiple we ask the user which to use.
+    if (opened.length === 1 || (opened.length > 1 && openRecent)) {
+      folderPath = opened[0].folderUri.path
+    } else if (opened.length > 1) {
+      const items: vscode.QuickPickItem[] = opened.map((folder): vscode.QuickPickItem => {
+        return {
+          label: folder.folderUri.path,
         }
-        selected = opened[items.indexOf(item)]
-      } else {
-        selected = opened[0]
+      })
+      const item = await vscode.window.showQuickPick(items, {
+        title: "Select a recently opened folder",
+      })
```

Comment:
Making sure I'm understanding this part: is the setup basically that we pass in a list of items to the function, it displays options, and then `item` becomes whichever one was selected?

Reply:
Honestly I am not sure this is the best way (it was already like this, I just shuffled it around), but yup `showQuickPick` returns the item that was selected (or undefined if canceled).

Reply:
Oh the part I was not sure about, is that we then use the index of that item to get the original item, I think it has to be done this way because we have to use a `QuickPickItem[]`, but I did not check to see if maybe there is a way to use a generic instead.

Reply:
Actually, since the label is the path, seems like we could just return that.  `item.label`.  :thinking:

Reply:
Refactored, looks much better!